packages/react-dom/src/shared/DOMProperty.js

```diff
@@ -197,85 +182,98 @@ export function shouldAttributeAcceptBooleanValue(name) {
-    // See http://schema.org/docs/gs.html
-    itemScope: HAS_BOOLEAN_VALUE,
-    // These attributes must stay in the white-list because they have
-    // different attribute names (see DOMAttributeNames below)
-    acceptCharset: 0,
-    className: 0,
-    htmlFor: 0,
-    httpEquiv: 0,
-    // Set the string boolean flag to allow the behavior
-    value: HAS_STRING_BOOLEAN_VALUE,
-  },
-  DOMAttributeNames: {
-    acceptCharset: 'accept-charset',
-    className: 'class',
-    htmlFor: 'for',
-    httpEquiv: 'http-equiv',
-  },
-};
+const attributeNames: Map<string, string> = new Map([
+  ['acceptCharset', 'accept-charset'],
+  ['className', 'class'],
+  ['htmlFor', 'for'],
+  ['httpEquiv', 'http-equiv'],
+
+  ['autoReverse', 'autoReverse'],
+  ['externalResourcesRequired', 'externalResourcesRequired'],
+  ['preserveAlpha', 'preserveAlpha'],
+]);
+
+export function getAttributeName(name: string): string {
+  let attributeName = attributeNames.get(name);
+  if (typeof attributeName === 'string') {
+    return attributeName;
+  }
+  attributeName = isWhitelisted(name) ? name.toLowerCase() : name;
+  attributeNames.set(name, attributeName);
+  return attributeName;
+}
+
+const NS_XLINK = 'http://www.w3.org/1999/xlink';
+const NS_XML = 'http://www.w3.org/XML/1998/namespace';
+const attributeNamespaces: Map<string, string> = new Map([
+  ['xlinkActuate', NS_XLINK],
+  ['xlinkArcrole', NS_XLINK],
+  ['xlinkHref', NS_XLINK],
+  ['xlinkRole', NS_XLINK],
+  ['xlinkShow', NS_XLINK],
+  ['xlinkTitle', NS_XLINK],
+  ['xlinkType', NS_XLINK],
+  ['xmlBase', NS_XML],
+  ['xmlLang', NS_XML],
+  ['xmlSpace', NS_XML],
+]);
+
+export function getAttributeNamespace(name: string): string | null {
+  const attributeNamespace = attributeNamespaces.get(name);
+  if (typeof attributeNamespace === 'string') {
+    return attributeNamespace;
+  }
+  return null;
+}
+
+export function hasBooleanValue(name: string): boolean {
+  return propertyTypes.get(name) === BOOLEAN;
+}
+
+export function hasOverloadedBooleanValue(name: string): boolean {
+  return propertyTypes.get(name) === OVERLOADED_BOOLEAN;
+}
 
-var NS = {
-  xlink: 'http://www.w3.org/1999/xlink',
-  xml: 'http://www.w3.org/XML/1998/namespace',
-};
+export function shouldIgnoreValue(name: string, value: mixed): boolean {
+  if (value == null) {
+    return true;
+  }
+  if (!propertyTypes.has(name)) {
+    return false;
+  }
+  const propertyType = propertyTypes.get(name);
+  switch (propertyType) {
+    case BOOLEAN:
+      return !value;
+    case OVERLOADED_BOOLEAN:
+      return value === false;
+    case POSITIVE_NUMERIC:
+      // Intentional implicit coercion (e.g. '0' < 1)
+      if ((value: any) < 1) {
+        return true;
+      }
+    // intentional fallthrough
+    case NUMERIC:
+      return isNaN(value);
+    default:
+      return false;
+  }
+}
+
+const usePropertiesFor: Set<string> = new Set([
+  'checked',
+  'multiple',
+  'muted',
+  'selected',
+]);
```

Comment:
I know microbenchmarks are dangerous, but would indexOf or object property membership be faster?

https://esbench.com/bench/5a21e22199634800a03494ec

<img width="1280" alt="screen shot 2017-12-01 at 6 14 23 pm" src="https://user-images.githubusercontent.com/590904/33507580-7b7c9918-d6c3-11e7-8c2a-e44805eacff4.png">

Reply:
I bet in the long-term checking set membership will be faster, since that's the sole purpose of the data structure. If we're still getting 25 million ops/s with a set it's probably fine to use it since its the correct data structure for this.

Reply:
Ya. I hesitated to say anything. Dan just mentioned the extra look-ups were slowing things down a bit.

Feel free to disregard.

Reply:
If it's a non-trivial performance regression we should definitely consider changing it. It's just so nice using `Set` and `Map` instead of arrays and objects 😍